File: app/websockets/binance_client.py
import asyncio

import websockets
import orjson

from app.registry.strategy_registry import (
    StrategyRegistry
)

from app.core.logger import logger
from app.core.config import settings


class BinanceWebSocketClient:

    registry = StrategyRegistry()

    symbols = (
        registry.get_symbols()
    )

    stream_url = (
        f"{settings.BINANCE_WS_BASE_URL}/stream?streams="
        + "/".join(
            [
                f"{symbol.lower()}@trade"
                for symbol in symbols
            ]
        )
    )

    async def connect(self):
        """
        Async generator that yields trade ticks forever.
        Reconnects automatically on any disconnect or timeout —
        using a manual retry loop instead of @retry so that it
        works correctly with async generators (tenacity @retry
        does not compose properly with `async def ... yield`).
        """

        reconnect_delay = 5  # seconds between reconnect attempts

        while True:

            try:

                logger.info(
                    f"Connecting to Binance: {self.stream_url}"
                )

                async with websockets.connect(
                    self.stream_url,
                    ping_interval=20,      # send a ping every 20s
                    ping_timeout=30,       # wait up to 30s for pong
                    close_timeout=10,
                ) as websocket:

                    logger.info("Connected to Binance WebSocket")
                    reconnect_delay = 5    # reset backoff on success

                    while True:

                        message = await websocket.recv()

                        payload = orjson.loads(message)

                        yield payload["data"]

            except (
                websockets.exceptions.ConnectionClosedError,
                websockets.exceptions.ConnectionClosedOK,
                asyncio.TimeoutError,
                OSError,
            ) as exc:

                logger.warning(
                    f"Binance WebSocket disconnected: {exc}. "
                    f"Reconnecting in {reconnect_delay}s..."
                )

                await asyncio.sleep(reconnect_delay)

                # Exponential backoff capped at 60s
                reconnect_delay = min(reconnect_delay * 2, 60)

            except Exception as exc:

                logger.error(
                    f"Unexpected stream error: {exc}. "
                    f"Reconnecting in {reconnect_delay}s..."
                )

                await asyncio.sleep(reconnect_delay)

                reconnect_delay = min(reconnect_delay * 2, 60)

File: backend/src/websockets/binanceClient.ts
import WebSocket from "ws";
import { settings } from "../core/config";
import { logger } from "../core/logger";
import { StrategyRegistry } from "../trading/strategies/strategyRegistry";

export class BinanceWebSocketClient {
  private registry = new StrategyRegistry();
  private symbols = this.registry.getSymbols();
  private streamUrl = `${settings.BINANCE_WS_BASE_URL}/stream?streams=` +
    this.symbols.map((symbol) => `${symbol.toLowerCase()}@trade`).join("/");

  private ws: WebSocket | null = null;
  private onTickCallback: ((tick: any) => Promise<void>) | null = null;

  constructor() {}

  connect(onTick: (tick: any) => Promise<void>) {
    this.onTickCallback = onTick;
    this.startConnection();
  }

  private startConnection() {
    let reconnectDelay = 5000; // milliseconds

    logger.info(`Connecting to Binance WebSocket: ${this.streamUrl}`);
    this.ws = new WebSocket(this.streamUrl);

    this.ws.on("open", () => {
      logger.info("Connected to Binance WebSocket");
      reconnectDelay = 5000; // reset delay
    });

    this.ws.on("message", async (data) => {
      try {
        const payload = JSON.parse(data.toString());
        if (payload && payload.data && this.onTickCallback) {
          await this.onTickCallback(payload.data);
        }
      } catch (err: any) {
        logger.error(`Error parsing message from Binance: ${err.message}`);
      }
    });

    this.ws.on("close", (code, reason) => {
      logger.warn(`Binance WebSocket closed (${code}): ${reason}. Reconnecting in ${reconnectDelay / 1000}s...`);
      this.reconnect(reconnectDelay);
    });

    this.ws.on("error", (err) => {
      logger.error(`Binance WebSocket error: ${err.message}`);
      this.ws?.close();
    });
  }

  private reconnect(delay: number) {
    const nextDelay = Math.min(delay * 2, 60000); // capped at 60s
    setTimeout(() => {
      this.startConnection();
    }, delay);
  }
}
export default BinanceWebSocketClient;
